Mobile: Fixes #12823: Limit how many search results are returned

pedr opened the pull request:
Fixes https://github.com/laurent22/joplin/issues/12823

# Summary

I'm limiting to 100 how many results can be found using the mobile search. This is a strategy to try to improve the user experience of users that have too many notes.

The limit is only set when the value is defined, so it should only affect the mobile app.

# Testing

I added some automated tests for the feature, including the search API.

Manually I tested by creating 4000+- notes by randomly selecting some lorem ipsum paragraphs. I tested by using keywords that would return 1500+- notes, using the performance logging utility I saw a drop from 1.5~2s to 0.3ms of response by limiting how many notes were returned.

While testing I also observed that the `Note.previews()` is probably the most expensive step when there is so many notes being returned

personalizedrefrigerator commented on packages/lib/services/rest/routes/search.test.ts:
```diff
@@ -0,0 +1,68 @@
+		}
+		await Promise.all(promises);
+		searchEngineInstance = SearchEngine.instance();
+		searchEngineInstance.setDb(db());
```
Thank you for working on this!

It could make sense to move `SearchEngine.instance().setDb` to [switchClient in `test-utils.ts`](https://github.com/laurent22/joplin/blob/0b6a1c75ba87754781274663283d22802693c2a5/packages/lib/testing/test-utils.ts#L289). This would reduce the setup logic included in `search.test.ts` and should make writing other tests for `SearchEngine` simpler in the future.

personalizedrefrigerator commented on packages/lib/services/rest/routes/search.test.ts:
```diff
@@ -0,0 +1,68 @@
+		for (let i = 0; i < 101; i++) {
+			promises.push(Note.save({ title: `abcd ${i}`, body: 'body' }));
+		}
+		await Promise.all(promises);
```
Since this "save 101 notes" logic is used in multiple places, consider moving it to a helper function.

personalizedrefrigerator commented on packages/app-mobile/components/screens/SearchScreen/SearchResults.tsx:
```diff
@@ -18,6 +18,8 @@ interface Props {
 	ftsEnabled: number;
 }
 
+const limit = 100;
```
**Possible UI adjustment**: To avoid confusion related to this behavior change, it may be helpful to show a "only the first ${limit} results are shown" message if the number of search results is equal to the limit.

pedr replied:
Good idea, thanks!

laurent22 commented on packages/lib/services/search/SearchEngine.ts:
```diff
@@ -767,6 +769,7 @@ export default class SearchEngine {
 			searchType: SearchEngine.SEARCH_TYPE_AUTO,
 			appendWildCards: false,
 			includeOrphanedResources: false,
+			limit: undefined,
```
Any reason you're setting it here when it's actually optional?

pedr replied:
I thought the pattern was to set the default value at the start of the function to make it visible.

laurent22 replied:
No there's no pattern to set properties to undefined

laurent22 commented on packages/lib/services/search/SearchEngineUtils.ts:
```diff
@@ -5,12 +5,14 @@ import Setting from '../../models/Setting';
 	public static async notesForQuery(query: string, applyUserSettings: boolean, options: NotesForQueryOptions = null, searchEngine: SearchEngine = null) {
 		options = {
 			appendWildCards: false,
+			limit: undefined,
```
same here, why set a variable to undefined when it's optional?

laurent22 commented on packages/lib/services/search/queryBuilder.ts:
```diff
@@ -444,7 +444,16 @@ const getConnective = (terms: Term[], relation: Relation): string => {
+
+export default function queryBuilder(terms: Term[], useFts: boolean, options: QueryBuilderOptions) {
+	options = {
+		limit: undefined,
```
same

laurent22 commented on packages/lib/services/rest/routes/search.test.ts:
```diff
@@ -0,0 +1,68 @@
+		await searchEngineInstance.syncTables();
+		await ItemChange.waitForAllSaved();
+
+		const result = await api.route(RequestMethod.GET, 'search', { query: 'abcd', limit: 100 });
+		const result2 = await api.route(RequestMethod.GET, 'search', { query: 'abcd', limit: 100, page: 2 });
```
I don't get why you're doing your tests with a limit of 100, which means you have to create more than one hundred notes? Is there something hard-coded somewhere that forces you to use 100 notes?

pedr replied:
My idea was to test that the limit used for mobile is not breaking anything in the API. Since in the mobile limit is  I'm testing here with 101 notes to ensure that is possible to get 101 of them.

The API uses the same method from `SearchEngineUtils.notesForQuery`

laurent22 commented on packages/lib/services/rest/routes/search.test.ts:
```diff
@@ -0,0 +1,68 @@
+		// Using 101 because the default limit is 100 returned results on mobile
+		for (let i = 0; i < 101; i++) {
+			promises.push(Note.save({ title: `abcd ${i}`, body: 'body' }));
+		}
```
Same, why 100?

pedr commented:
I added an automated test to the mobile component and added images to the PR showing how the limit message is being shown.

personalizedrefrigerator commented on packages/app-mobile/components/screens/SearchScreen/SearchResults.tsx:
```diff
@@ -113,10 +155,19 @@ const SearchResults: React.FC<Props> = props => {
 				keyExtractor={(item) => item.id}
-				renderItem={event => <NoteItem note={event.item} />}
+				renderItem={event => {
+					if (event.index === 0 && notes.length === limit) {
+						return <LimitMessage themeId={props.themeId} />;
+					}
```
Will this prevent the first result from being rendered?

pedr commented:
As Henry pointed out there was a bug where the first result wouldn't show up. I fixed the bug and was trying to add a test for it by using `fireEvent.scroll` to check how many items there is in the list, but I couldn't find the proper way to scroll and wait, so I end up going with the simplest implementation to be able to add a test that is adding a new prop that controls how many items are being rendered from the Flatlist (by default is 10 and it should still be since the value is undefined everywhere the component is called but the test that I added).

personalizedrefrigerator commented on packages/app-mobile/components/screens/SearchScreen/SearchResults.tsx:
```diff
@@ -113,10 +156,23 @@ const SearchResults: React.FC<Props> = props => {
+				initialNumToRender={props.initialNumToRender}
+				renderItem={event => {
+					if (event.index === 0 && notes.length === limit) {
+						return <React.Fragment>
+							<LimitMessage themeId={props.themeId} />
+							<NoteItem note={event.item} />
+						</React.Fragment>;
```
**Refactoring/non-blocking**: Consider using the [`ListHeaderComponent`](https://reactnative.dev/docs/flatlist#listheadercomponent) prop to display the limit message, rather than including the message in the first result item.

personalizedrefrigerator commented:
**Note**: Since Joplin's [search sorting logic](https://github.com/laurent22/joplin/blob/dfbd5eb8eda61f1df9d461f722fb7a40beb592f0/packages/lib/services/search/SearchEngine.ts#L505) currently runs after querying all results (rather than as a part of the SQL query), the approach taken here results in search not showing the [BM25 "most relevant"](https://github.com/laurent22/joplin/blob/dfbd5eb8eda61f1df9d461f722fb7a40beb592f0/packages/lib/services/search/SearchEngine.ts#L370) results on top.

laurent22 commented:
Fixed in #13444
